### PR TITLE
fix: Allow patches with empty files with multiple newlines or comments

### PR DESCRIPTION
**Pull request by jchanam**
Fixes https://github.com/kubernetes-sigs/kustomize/issues/5487

Right now, when a file is empty, the patches given by the resmap (sm patches) or the jsonpatch is nil. When the file has only comments in it (virtually empty if we think about changes to be made) or multiple newlines, the result given by said functions is an empty slice.

I've changed the check for when we can't allow to have a patch that is valid for both SM and JSON, so it allows empty changes (both nil or empty slices).

Extra comments:
- I think that it'd be better to change the functions and return a nil object, instead of an empty slice, but I'm unsure on where to do those changes.
- Also, I can't make the tests to run locally as they complain for the content file to be empty, and that should not be the case as it's perfectly ok for the built binary to use empty patches.

I'll be more than happy to work on these two with some help or guidance.

### CODE REVIEW

**Comment by stormqueen1990** [comment]
> Hi @stormqueen1990 @varshaprasad96,
> 
> Could I get some help in getting this approved?
> 
> Thank you!

Hi there, @jchanam!

I am taking a look at this and got a bit confused with the comment you made in the PR description about the tests. Do you mean that tests are currently failing locally?

**Comment by jchanam** [comment]
Hi @stormqueen1990 ,

Yes, I meant that. It looks like there are a test checking if the input is not empty, but the goal of my code is testing specifically that, as it should not fail. If we add a file to kustomize that is empty, it should simply not produce any output, instead of failing.

Also, now the tests here say that there's a diff found on the file I've modified. How can I fix that or change extra code so the tests continue?

Thank you!

File: plugin/builtin/patchtransformer/PatchTransformer_test.go [hunk 1]
```diff
@@ -1108,3 +1108,32 @@ spec:
...
 `)
 }
+
+func TestPatchTransformerPatchEmpty(t *testing.T) {
+	th := kusttest_test.MakeEnhancedHarness(t).
+		PrepBuiltin("PatchTransformer")
+	defer th.Reset()
+
+	th.RunTransformerAndCheckError(`
+`, someDeploymentResources, func(t *testing.T, err error) {
```
Comment (jchanam) [inline plugin/builtin/patchtransformer/PatchTransformer_test.go:1118]: @stormqueen1990 This is an example of what I mean.

Here I'm testing the input to be a file with only a new line. Based on my new code, this should give no error, but I'm getting a specific error from the tests extra code saying that the input to RunTransformerAndCheckError can't be empty.

Comment (stormqueen1990) [reply plugin/builtin/patchtransformer/PatchTransformer_test.go:1118]: You're missing a `go generate` step for this code. The `plugin/builtin/patchtransformer/PatchTransformer.go` file you updated is a scaffold for the plugin generated under `api/internal/builtins/PatchTransformer.go`; therefore, you will need to run `go generate plugin/builtin/patchtransformer/PatchTransformer.go` to update the final built-in plugin code.

You can also run the prow checks locally using `make prow-presubmit-check` from the root of the repository.

Comment (stormqueen1990) [reply plugin/builtin/patchtransformer/PatchTransformer_test.go:1118]: Hi there, @jchanam! Sorry for the delay in responding. Let me take a look at this and I'll get back to you!

Comment (stormqueen1990) [reply plugin/builtin/patchtransformer/PatchTransformer_test.go:1118]: Hi again, @jchanam!

I took a deeper look and this test case isn't quite right. If you copy it and paste in a Kustomize version without the regression (e.g. tag `kustomize/v5.1.1`) it still won't pass.

Taking a look at the configuration for `RunTransformerAndCheckError`, the first argument of the function refers to the transformer configuration and that cannot be empty under any circumstance:
https://github.com/kubernetes-sigs/kustomize/blob/7911b2c001a144e5ccef3026ec2c48a94140423c/api/testutils/kusttest/harnessenhanced.go#L200-L204

The transformer configuration is the blurb that instructs Kustomize how to deal with a patch. In the `PatchTransformer` case, that would be this part of the `kustomization.yaml` file:
```yaml
patches:
  - target:
      group: apps
      version: v1
      kind: Deployment
      name: test-deployment
    path: patches/test-patch.yaml
```

What we want to be empty in this scenario is the patch, i.e. the file to which the transformer configuration refers, not the transformer configuration. The tests that write patch files in this test file are more similar to the scenario you need to emulate (for example, `TestPatchTransformerSmpSidecars`).

**Comment by jchanam** [comment]
@stormqueen1990 Sorry for the late response. I think I was able to follow your guidance and fix the tests.

File: plugin/builtin/patchtransformer/PatchTransformer_test.go (hunk 1 above)
Comment (jchanam) [reply plugin/builtin/patchtransformer/PatchTransformer_test.go:1128]: Hi @stormqueen1990 !

Thank you for the help! I was able to fix the tests and make them work 👯‍♂️

**Comment by stormqueen1990** [comment]
> @stormqueen1990 Sorry for the late response. I think I was able to follow your guidance and fix the tests.

Hi there, @jchanam! Thanks for updating this!

**Comment by jchanam** [comment]
@stormqueen1990 looks like the tests were not run. Could you please force them in any way? thank you!

**Comment by stormqueen1990** [comment]
> @stormqueen1990 Could you please help me with the linter errors?
> 
> ```
> Error: plugin/builtin/patchtransformer/PatchTransformer_test.go:1128:29: test helper function should start from t.Helper() (thelper)
> `, someDeploymentResources, func(t *testing.T, err error) {
>                             ^
> Error: plugin/builtin/patchtransformer/PatchTransformer_test.go:1154:29: test helper function should start from t.Helper() (thelper)
> `, someDeploymentResources, func(t *testing.T, err error) {
> ```
> 
> I don't understand why it's complaining when I've followed the same structure as for other tests

Hi @jchanam, our linter requires test helper functions to have an invocation to `t.Helper()` at the beginning. You just need to add that statement as the first statement in the helper function. This is standard Go, you can read more about it here: https://pkg.go.dev/testing#T.Helper. I apologize for not seeing (and flagging) it earlier.

As for the tests, `/test all` and `/retest` will only trigger the Prow presubmit tests. You can check the logs for those tests by clicking the details in the `kustomize-presubmit-master` check in your pull request status.

**Review by stormqueen1990 — COMMENTED** [review]
/lgtm

**Comment by jchanam** [comment]
@stormqueen1990 what is the following step now that the PR passed all the tests? Could this PR be merged?

**Comment by stormqueen1990** [comment]
> @stormqueen1990 what is the following step now that the PR passed all the tests? Could this PR be merged?

Hi there, @jchanam! The next step is the review and then approval by a maintainer (I am just a reviewer). Once one of the maintainers approve this pull request using `/approve` it will be merged automatically by Prow. You can read more about the pull request process in Kubernetes projects in the Contributor Documentation: https://www.kubernetes.dev/docs/guide/pull-requests/

**Comment by jchanam** [comment]
Hi!

I'm still waiting for any maintainer to approve and merge this. Is there any way to ping anyone so they can check this?

**Comment by stormqueen1990** [comment]
> Hi!
> 
> I'm still waiting for any maintainer to approve and merge this. Is there any way to ping anyone so they can check this?

@koba1t or @varshaprasad96: could you please take a look at this PR when you get a chance?

**Comment by jchanam** [comment]
Hi! Could we have this PR merged please?

**Comment by koba1t** [comment]
Thanks @jchanam 

/lgtm
/approve
